### PR TITLE
Add possibility to mapCube over moving windows

### PR DESCRIPTION
**Pull request by meggart**
This is a first implementation of Moving windows in mapCube. @felixcremer if you have time to try this out, please let me know if/how it works. A minimal example would be:

````julia
using YAXArrays
#Create a dummy array
a = Array{Union{Float64,Missing}}(rand(40,20,10))
lon = RangeAxis("Lon",1:40)
lat = RangeAxis("Lat",1:20)
tim = RangeAxis("Time",1:10)
c = YAXArray([lon,lat,tim],a)

#Define the input dimensions with time first (all time steps) and lon and lat each with one previous and one consecutive slice. 
indims = InDims("Time",MovingWindow("Lon",1,1),MovingWindow("Lat",1,1))

r1 = mapCube(c, indims=indims, outdims=OutDims("Time")) do xout,xin
    #Inside this function, xin will have size 10x3x3 (time x lon x lat)
end
````

### CODE REVIEW

**Review by felixcremer — COMMENTED** [review]
This looks good to me. We need to write some documentation for this moving window functionality. Maybe I can write down something next Wednesday. 
And I would also like to test this on larger cubes, but this is non-blocking.

**Comment by felixcremer** [comment]
The moving window approach fails, when I use it with a datacube which doesn't allow missings and I am doing something where I insert missings, at the border, because I am trying to access stuff from outside my datacube. 
Wouldn't it be nicer, when the outcube eltype would be widened to allow for missings?

This is basically the same data as in the unit tests, but with an input data array of eltype Float64 instead of Union{Missing, Float64}
```julia
julia> r1 = mapCube(c, indims=indims, outdims=OutDims("Time")) do xout,xin
       xout[:] = xin[:,1]
       end
Error during running loop: MethodError(convert, (Float64, missing), 0x0000000000006cf4)
ERROR: MethodError: Cannot `convert` an object of type Missing to an object of type Float64
Closest candidates are:
  convert(::Type{T}, ::Ratios.SimpleRatio{S}) where {T<:AbstractFloat, S} at /home/crem_fe/.julia/packages/Ratios/uRs4y/src/Ratios.jl:13
  convert(::Type{T}, ::ArrayInterface.StaticInt{N}) where {T<:Number, N} at /home/crem_fe/.julia/packages/ArrayInterface/gMtB5/src/static.jl:22
  convert(::Type{T}, ::YAXArrays.DAT.SentinelMissings.SentinelMissing) where T<:Number at /home/crem_fe/.julia/dev/YAXArrays/src/DAT/SentinelMissings.jl:9
  ...
Stacktrace:
 [1] runLoop(::YAXArrays.DAT.DATConfig{1,1}, ::Bool) at /home/crem_fe/.julia/dev/YAXArrays/src/DAT/DAT.jl:492
 [2] mapCube(::var"#3#4", ::Tuple{YAXArray{Float64,3,Array{Float64,3},Array{RangeAxis{Int64,S,UnitRange{Int64}} where S,1}}}; max_cache::Float64, indims::InDims, outdims::OutDims, inplace::Bool, ispar::Bool, debug::Bool, include_loopvars::Bool, showprog::Bool, nthreads::Array{Int64,1}, loopchunksize::Dict{Any,Any}, kwargs::Base.Iterators.Pairs{Union{},Union{},Tuple{},NamedTuple{(),Tuple{}}}) at /home/crem_fe/.julia/dev/YAXArrays/src/DAT/DAT.jl:310
 [3] #mapCube#42 at /home/crem_fe/.julia/dev/YAXArrays/src/DAT/DAT.jl:209 [inlined]
 [4] top-level scope at REPL[15]:1
```

File: Project.toml [hunk 1]
```diff
@@ -40,6 +41,7 @@ Glob = "1.3"
 Interpolations = "0.12, 0.13"
 IntervalSets = "0.3, 0.4, 0.5"
 IterTools = "1"
+OffsetArrays = "1.5"
```
[the remark this answers is not in the document]
Comment (meggart) [reply Project.toml:44]: Ok, I changed this

File: src/DAT/DAT.jl [hunk 2]
```diff
@@ -31,26 +35,44 @@ Internal representation of an input cube for DAT operations
...
+  fullaxes = internalaxes[findall(i->!isa(i,MovingWindow), desc.axisdesc)]
   axlist = caxes(c)
-  axesSmall = map(i->axlist[i], iaxessmall)
-  colonperm = issorted(iaxessmall) ? nothing : collect(Base.invperm(sortperm(collect(iaxessmall))))
-  iaxessmall = collect(iaxessmall)
-  any(isequal(nothing),axesSmall) && error("One of the input axes not found in input cubes")
-  InputCube{ndims(c)}(c,desc,collect(CubeAxis,axesSmall),iaxessmall,colonperm,Int[],Int[])
+  axesSmall = map(i->axlist[i], fullaxes)
+  colonperm = issorted(internalaxes) ? nothing : collect(Base.invperm(sortperm(collect(internalaxes))))
+  iaxessmall = collect(fullaxes)
```
[the remark this answers is not in the document]
Comment (meggart) [reply src/DAT/DAT.jl:56]: done

File: src/DAT/registration.jl [hunk 3]
```diff
@@ -46,27 +57,31 @@ getprocfilter(pf::ProcFilter) = (pf,)
...
 
 ### Keyword arguments
 
-* `artype` how shall the array be represented in the inner function. Defaults to `AsArray`, alternatives are `AsDataFrame` or `AsAxisArray`
+* `artype` how shall the array be represented in the inner function. Defaults to `Array`, alternatives are `DataFrame` or `AsAxisArray`
 * `filter` define some filter to skip the computation, e.g. when all values are missing. Defaults to
     `AllMissing()`, possible values are `AnyMissing()`, `AnyOcean()`, `StdZero()`, `NValid(n)`
     (for at least n non-missing elements). It is also possible to provide a custom one-argument function
     that takes the array and returns `true` if the compuation shall be skipped and `false` otherwise.
+* `window` key-value iterable of the form `dimname => (pre, after)` describing if the mapping operation shall be
```
[the remark this answers is not in the document]
Comment (meggart) [reply src/DAT/registration.jl:75]: Thanks for the catch, this was my initial experiment

File: src/YAXTools.jl [hunk 4]
```diff
@@ -2,54 +2,71 @@ module YAXTools
...
 import ..YAXArrays: YAXdir
 export freshworkermodule, passobj, @everywhereelsem,
-@loadOrGenerate, PickAxisArray
+@loadOrGenerate, PickAxisArray, Window
 struct PickAxisArray{T,N,AT<:AbstractArray,P,PERM}
-    parent::AT
+  parent::AT
 end
 
+struct Window
```
[the remark this answers is not in the document]
Comment (meggart) [reply src/YAXTools.jl:10]: Good question. Window is basically what PickAxisArray is using to represent these windows, while MovingWindow is a descirption used by InputCube and also contains a notion of axis names. I think in general PickAxisArrays might live in a separate package, becuase it represents a very raw concept not tied to data cubes, so I think it is quite natural that both use their own window definition type.

**Comment by meggart** [comment]
? The moving window approach fails, when I use it with a datacube which doesn't allow missings and I am doing something where I insert missings, at the border, because I am trying to access stuff from outside my datacube.
Wouldn't it be nicer, when the outcube eltype would be widened to allow for missings?

I did not really want to widen the output array type if not necessary, so I added a keyword argument to `InDims` to let the user decide which value to fill the oob areas with. The tests work now when setting `indims = InDims("Time",MovingWindow("Lon",1,1), window_oob_value = -9999.0)` and the areas will be filled with the respective value.

**Comment by meggart** [comment]
I will merge this for now and start on cleaning up the documentation for this package. Maybe we can move some parts from ESDL.jl here, but probably it would make more sense to just start over and fill piece by piece.

**Comment by dpabon** [comment]
The window_oob_value didn't resolve the issue when data is Float32, or when two moving windows are used at the same time? (e.g. lat and lon) not sure which one is correct:

here, a minimum example:

```julia
using EarthDataLab
using NetCDF
using YAXArrays
using DataFrames
using XLSX
using CSV
using Zarr
using Plots
using CairoMakie
using GeoMakie
using Statistics
# Dataproduct
sim_cube_path = "/Net/Groups/BGI/people/awinkler/share/Historical-LandUse_CMIP6.nc"

sim_cube = open_dataset(sim_cube_path)

sim_cube

sim_cube = Cube(sim_cube)


# Checking that winsize is odd

if isodd(winsize)
    pre_step = after_step = floor(winsize / 2)
    else
        pre_step = after_step = floor(winsize / 2) -1

        @warn "Window size is not odd. Going on however... windowsize = $(winsize - 1)"
end

# Define co-ocurrence function


q1 = [0.3 0.3 0.3 ; 0.3 0.3 0.3 ; 0.4 0.4 0.4]

q2 = [0.1 missing 0.3 ; 0.5 0.1 0.3 ; 0.6 0.4 0.4]


 #q1 = [0.3, 0.4, 0.5]
 #q2 = [0.3, 0.2, 0.1]

coocufun = function (out, q1, q2)
    replace!(q1, NaN => missing)
    replace!(q2, NaN => missing)
    # Where q1 and q2 are the ratio of presence of pft1 and pft2 in the moving window.
     #@show typeof(q1), typeof(q2)
     #println(q1, q2)

    if length(q1) == length(q2) && (!any(ismissing.(q1))) && !any(ismissing.(q2))
        #q1 = vec(q1)
        #q2 = vec(q2)
         p1 = range(0, 1, length = length(q1))
         p2 = reverse(p1)
         pmindist = minimum(eachrow(([i - j for i in q1, j in p1]).^2 + ([i - j for i in q2, j in p2]).^2))
         denom = sum(sqrt.(sum.(eachrow([p1   p2].^2))))

         out[1] = 1 - (sum(sqrt.(pmindist)) / denom)

         #return 1 - (sum(sqrt.(pmindist)) / denom)
    else
         out[1] = 0 # 0 because assume not enough data on the moving window.
         #return 0

    end
end

# testing coocufun using the simulated datacube.


#Define the input dimensions with time first (all time steps) and lon and lat each with one previous and one consecutive slice. 

indims = InDims(MovingWindow("lon", pre_step,after_step),
MovingWindow("lat", pre_step, after_step), window_oob_value=NaN)

outdims = OutDims()

result_coocurrence = mapCube(coocufun, 
(sim_cube_small[variable = "grassFrac"], sim_cube_small[variable = "treeFrac"]), 
indims = (indims, indims), 
outdims = outdims)
``` 

produce:

```julia
nDims((MovingWindow(YAXArrays.Cubes.Axes.ByName("lon"), 2, 2), MovingWindow(YAXArrays.Cubes.Axes.ByName("lat"), 2, 2)), Array, (YAXArrays.DAT.AllMissing(),), NaN)

OutDims((), :auto, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}(), false, Array, :input, 1)

ERROR: MethodError: Cannot `convert` an object of type Missing to an object of type Float32
Closest candidates are:
  convert(::Type{T}, ::Gray24) where T<:Real at /Net/Groups/BGI/people/dpabon/bin/julia_packages/packages/ColorTypes/1dGw6/src/conversions.jl:114
  convert(::Type{T}, ::Gray) where T<:Real at /Net/Groups/BGI/people/dpabon/bin/julia_packages/packages/ColorTypes/1dGw6/src/conversions.jl:113
  convert(::Type{T}, ::DualNumbers.Dual) where T<:Union{Real, Complex} at /Net/Groups/BGI/people/dpabon/bin/julia_packages/packages/DualNumbers/5knFX/src/dual.jl:24
  ...
Stacktrace:
  [1] setindex!(A::Matrix{Float32}, x::Missing, i1::Int64)
    @ Base ./array.jl:903
  [2] _replace!(new::Base.var"#new#329"{Tuple{Pair{Float64, Missing}}}, res::Matrix{Float32}, A::Matrix{Float32}, count::Int64)
    @ Base ./set.jl:681
  [3] replace_pairs!
    @ ./set.jl:499 [inlined]
  [4] #replace!#328
    @ ./set.jl:489 [inlined]
  [5] replace!
    @ ./set.jl:489 [inlined]
  [6] (::var"#89#92")(out::Array{Union{Missing, Float32}, 0}, q1::Matrix{Float32}, q2::Matrix{Float32})
    @ Main ~/dpabon/nfdi4earth_oemc/bin/julia/exploratory_analysis/sim_cube.jl:144
  [7] innercode(f::Function, cI::CartesianIndex{4}, xinBC::Tuple{YAXArrays.YAXTools.PickAxisArray{Float32, 4, OffsetArrays.OffsetArray{Float32, 4, Array{Float32, 4}}, (YAXArrays.YAXTools.Window(1, 2, 2), YAXArrays.YAXTools.Window(2, 2, 2), 3, 4), nothing}, YAXArrays.YAXTools.PickAxisArray{Float32, 4, OffsetArrays.OffsetArray{Float32, 4, Array{Float32, 4}}, (YAXArrays.YAXTools.Window(1, 2, 2), YAXArrays.YAXTools.Window(2, 2, 2), 3, 4), nothing}}, xoutBC::Tuple{YAXArrays.YAXTools.PickAxisArray{Union{Missing, Float32}, 4, Array{Union{Missing, Float32}, 4}, (1, 2, 3, 4), nothing}}, filters::Tuple{Tuple{YAXArrays.DAT.AllMissing}, Tuple{YAXArrays.DAT.AllMissing}}, inwork::Tuple{Vector{Matrix{Float32}}, Vector{Matrix{Float32}}}, outwork::Tuple{Vector{Array{Union{Missing, Float32}, 0}}}, axvalcreator::YAXArrays.DAT.NoLoopAxes, offscur::NTuple{4, Int64}, addargs::Tuple{}, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ YAXArrays.DAT /Net/Groups/BGI/people/dpabon/bin/julia_packages/packages/YAXArrays/Ly8sc/src/DAT/DAT.jl:1075
  [8] innerLoop(loopRanges::NTuple{4, UnitRange{Int64}}, f::Function, xinBC::Tuple{YAXArrays.YAXTools.PickAxisArray{Float32, 4, OffsetArrays.OffsetArray{Float32, 4, Array{Float32, 4}}, (YAXArrays.YAXTools.Window(1, 2, 2), YAXArrays.YAXTools.Window(2, 2, 2), 3, 4), nothing}, YAXArrays.YAXTools.PickAxisArray{Float32, 4, OffsetArrays.OffsetArray{Float32, 4, Array{Float32, 4}}, (YAXArrays.YAXTools.Window(1, 2, 2), YAXArrays.YAXTools.Window(2, 2, 2), 3, 4), nothing}}, xoutBC::Tuple{YAXArrays.YAXTools.PickAxisArray{Union{Missing, Float32}, 4, Array{Union{Missing, Float32}, 4}, (1, 2, 3, 4), nothing}}, filters::Tuple{Tuple{YAXArrays.DAT.AllMissing}, Tuple{YAXArrays.DAT.AllMissing}}, inwork::Tuple{Vector{Matrix{Float32}}, Vector{Matrix{Float32}}}, outwork::Tuple{Vector{Array{Union{Missing, Float32}, 0}}}, axvalcreator::YAXArrays.DAT.NoLoopAxes, addargs::Tuple{}, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ YAXArrays.DAT /Net/Groups/BGI/people/dpabon/bin/julia_packages/packages/YAXArrays/Ly8sc/src/DAT/DAT.jl:1098
  [9] (::YAXArrays.DAT.var"#108#112"{YAXArrays.DAT.DATConfig{2, 1}})(r::NTuple{4, UnitRange{Int64}})
    @ YAXArrays.DAT /Net/Groups/BGI/people/dpabon/bin/julia_packages/packages/YAXArrays/Ly8sc/src/DAT/DAT.jl:663
 [10] (::ProgressMeter.var"#56#59"{Distributed.RemoteChannel{Channel{Bool}}, YAXArrays.DAT.var"#108#112"{YAXArrays.DAT.DATConfig{2, 1}}})(x::NTuple{4, UnitRange{Int64}})
    @ ProgressMeter /Net/Groups/BGI/people/dpabon/bin/julia_packages/packages/ProgressMeter/sN2xr/src/ProgressMeter.jl:1016
 [11] iterate
    @ ./generator.jl:47 [inlined]
 [12] _collect(c::DiskArrays.GridChunks{4}, itr::Base.Generator{DiskArrays.GridChunks{4}, ProgressMeter.var"#56#59"{Distributed.RemoteChannel{Channel{Bool}}, YAXArrays.DAT.var"#108#112"{YAXArrays.DAT.DATConfig{2, 1}}}}, #unused#::Base.EltypeUnknown, isz::Base.HasShape{4})
    @ Base ./array.jl:744
 [13] collect_similar(cont::DiskArrays.GridChunks{4}, itr::Base.Generator{DiskArrays.GridChunks{4}, ProgressMeter.var"#56#59"{Distributed.RemoteChannel{Channel{Bool}}, YAXArrays.DAT.var"#108#112"{YAXArrays.DAT.DATConfig{2, 1}}}})
    @ Base ./array.jl:653
 [14] map(f::Function, A::DiskArrays.GridChunks{4})
    @ Base ./abstractarray.jl:2867
 [15] macro expansion
    @ /Net/Groups/BGI/people/dpabon/bin/julia_packages/packages/ProgressMeter/sN2xr/src/ProgressMeter.jl:1015 [inlined]
 [16] macro expansion
    @ ./task.jl:399 [inlined]
 [17] macro expansion
    @ /Net/Groups/BGI/people/dpabon/bin/julia_packages/packages/ProgressMeter/sN2xr/src/ProgressMeter.jl:1014 [inlined]
 [18] macro expansion
    @ ./task.jl:399 [inlined]
 [19] progress_map(::Function, ::Vararg{Any}; mapfun::Function, progress::ProgressMeter.Progress, channel_bufflen::Int64, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ ProgressMeter /Net/Groups/BGI/people/dpabon/bin/julia_packages/packages/ProgressMeter/sN2xr/src/ProgressMeter.jl:1007
 [20] progress_map(::Function, ::Vararg{Any})
    @ ProgressMeter /Net/Groups/BGI/people/dpabon/bin/julia_packages/packages/ProgressMeter/sN2xr/src/ProgressMeter.jl:1003
 [21] runLoop(dc::YAXArrays.DAT.DATConfig{2, 1}, showprog::Bool)
    @ YAXArrays.DAT /Net/Groups/BGI/people/dpabon/bin/julia_packages/packages/YAXArrays/Ly8sc/src/DAT/DAT.jl:661
 [22] mapCube(::var"#89#92", ::Tuple{YAXArray{Float32, 4, DiskArrays.SubDiskArray{Float32, 4}, Vector{CubeAxis}}, YAXArray{Float32, 4, DiskArrays.SubDiskArray{Float32, 4}, Vector{CubeAxis}}}; max_cache::Float64, indims::Tuple{InDims, InDims}, outdims::OutDims, inplace::Bool, ispar::Bool, debug::Bool, include_loopvars::Bool, showprog::Bool, irregular_loopranges::Bool, nthreads::Vector{Int64}, loopchunksize::Dict{Any, Any}, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ YAXArrays.DAT /Net/Groups/BGI/people/dpabon/bin/julia_packages/packages/YAXArrays/Ly8sc/src/DAT/DAT.jl:444
 [23] top-level scope
    @ ~/dpabon/nfdi4earth_oemc/bin/julia/exploratory_analysis/sim_cube.jl:186
```

**Pull request merged**